numpy/core/src/multiarray/refcount.c

```diff
@@ -51,6 +51,31 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
             PyArray_Item_INCREF(data + offset, new);
         }
     }
+    else if (PyDataType_HASSUBARRAY(descr)) {
+        int size, i, inner_elsize;
+
+        inner_elsize = descr->subarray->base->elsize;
+        /* Subarrays are always contiguous in memory */
+        size = descr->elsize / inner_elsize;
+
+        if (descr->subarray->base->type_num == NPY_OBJECT) {
+            /* Fast path for only-objects */
+            for (i = 0; i < size; i++){
+                NPY_COPY_PYOBJECT_PTR(&temp, data + i*inner_elsize);
+                Py_XINCREF(temp);
+            }
+        }
+        else {
+            for (i = 0; i < size; i++){
+                PyArray_Item_INCREF(data + i * inner_elsize,
```

Comment:
OK, I think I finally understand what this is actually doing - it just checks whether any part is an object and then increases its refcount. It might not be bad to write a comment to state that, so that the next person looking at the code doesn't have to look up what `PyArray_Item_INCREF` actually does - the name is not too helpful here.

Separately, the check for object is very early on; I don't think there is much to gain by making a fast path, especially since anything to do with objects is slow anyway.

Reply:
I was not 100% sure where you were going. Added some documentation in a new commit though.